### PR TITLE
authorize update entry to author only

### PR DESCRIPTION
**Pull request by manandearth**
## Goal
Authorize update of entries by `:authored`
So that some routes are permitted to author and admin only.
## Proposed changes
- `ALTER TABLE users ADD COLUMN authored VARCHAR(20);`
- Extend authorization depending on ownership (authorship) 
- add username to users table from `(get-in  request [session :identity :username])`

### CODE REVIEW

**Comment by manandearth** [comment]
Just read your review about the separate concerns (autorization and authentication) 
I wonder what you will make of this hybrid..

**Comment by vemv** [comment]
> Add The username to the table along with an entry .

Sorry, I don't understand this. Explain the actual goal: the why, not the how

**Comment by manandearth** [comment]
Having merged with `authorizing` branch, We have some working routes. 
I'll branch out again for unit-tests.

File: src/playground/services/invoices/insert/endpoint.clj [hunk 1]
```diff
@@ -8,9 +8,11 @@
 
 (spec/def ::api (spec/keys :req-un [::models.user/amount]))
 
-(defn perform [{{:keys [amount]} :form-params :keys [db] :as request}]
+(defn perform [{{:keys [amount]} :form-params :keys [db session] :as request}]
   (let [db     (->> db :pool (hash-map :datasource))
-        insert (-> (logic/to-insert amount (java.util.UUID/randomUUID))
+        insert (-> (logic/to-insert amount
+                                    (java.util.UUID/randomUUID)
+                                    session)
```
Comment (vemv) [inline src/playground/services/invoices/insert/endpoint.clj]: the session is an HTTP concept. Don't pass it to the `logic`. Instead, extract the username just here and pass it as an argument.

File: src/playground/services/invoices/retrieve/endpoint.clj [hunk 2]
```diff
@@ -27,3 +27,11 @@
     (if record
       (ring-resp/response (view/update-invoice request record))
       {:status 404 :body "Entry not in DB"})))
+
+(defn return-authored [{{:keys [id]} :path-params :keys [db] :as request}]
```
Comment (vemv) [inline src/playground/services/invoices/retrieve/endpoint.clj]: avoid the word "return". Most of defns return something. Better: `get-author`, `author-of`

File: README.md [hunk 3]
```diff
@@ -40,7 +40,7 @@ It requires to have the following namespaces required:
 
 For the service.invoices ns create a table:
 
-`CREATE TABLE users (id SERIAL, email VARCHAR(50))`
+`CREATE TABLE users (id SERIAL, email VARCHAR(50), authored VARCHAR(20))`
```
Comment (vemv) [inline README.md]: this name is a bit ambigous. Better: `author`, `authored-by`, `created-by` etc

**Review by vemv — COMMENTED** [review]
As a non-admin, I cannot delete a user I created

**Comment by manandearth** [comment]
> As a non-admin, I cannot delete a user I created

This "design" I am using in the parallel project, where every entry is a swallow/swift nest. the idea is that a nest can only be destroyed  (there's a boolean field for that) and therefore it cannot be deleted and the admin deletes entries in case there are duplicates and mistakes,

File: README.md (hunk 3 above)
Comment (manandearth) [reply README.md]: will change to author

File: src/playground/service.clj [hunk 4]
```diff
@@ -92,6 +95,25 @@
...
+(def author-interceptor
+  "throw unauthorized 403 by author
+  (allows admin and author of entry)"
+  {:name ::author-interceptor
+   :enter (fn [context]
+            (let [role     (get-in context [:request :session :identity :role])
+                  username (get-in context [:request :session :identity :username])
+                  id       (get-in context [:request :path-params :id])
+                  db       (get-in context [:request :db])
+                  author (:author (invoices.retrieve/get-author (:request context)))]
```
Comment (vemv) [inline src/playground/service.clj:107]: Different endpoints may have different ways of deriving its corresponding author.

But here you are using `invoices.retrieve` to derive the author for every single consumer of `author-interceptor`.

i.e. we tie a specific logic to the general behavior, which is wrong.

instead, _parameterize_ `author-interceptor`, i.e. make it a `defn` which accepts a `fn` as its argument:

image

**Review by vemv — APPROVED** [review]
Feedback for later

**Pull request merged**